src/Microsoft.TemplateEngine.Cli/Commands/InstantiateCommand.TabCompletion.cs

```diff
@@ -0,0 +1,74 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+#nullable enable
+
+using System.CommandLine.Parsing;
+using Microsoft.TemplateEngine.Abstractions;
+using Microsoft.TemplateEngine.Edge.Settings;
+
+namespace Microsoft.TemplateEngine.Cli.Commands
+{
+    internal partial class InstantiateCommand : BaseCommand<InstantiateCommandArgs>
+    {
+        internal IEnumerable<string> GetSuggestions(
+            InstantiateCommandArgs args,
+            IEnumerable<TemplateGroup> templateGroups,
+            IEngineEnvironmentSettings environmentSettings,
+            TemplatePackageManager templatePackageManager,
+            string? textToMatch)
+        {
+            if (string.IsNullOrWhiteSpace(args.ShortName))
+            {
+                return templateGroups.SelectMany(g => g.ShortNames).Distinct();
+            }
+            else
+            {
+                HashSet<string> distinctSuggestions = new HashSet<string>();
+                foreach (TemplateGroup templateGroup in templateGroups.Where(template => template.ShortNames.Contains(args.ShortName)))
+                {
+                    foreach (IGrouping<int, CliTemplateInfo> templateGrouping in templateGroup.Templates.GroupBy(g => g.Precedence).OrderByDescending(g => g.Key))
+                    {
+                        foreach (CliTemplateInfo template in templateGrouping)
+                        {
+                            TemplateCommand command = new TemplateCommand(this, environmentSettings, templatePackageManager, templateGroup, template);
+                            Parser parser = ParserFactory.CreateTemplateParser(command);
+                            ParseResult templateParseResult = parser.Parse(args.RemainingArguments ?? Array.Empty<string>());
+
+                            //TODO:
+                            // command.GetSuggestions does not work for arguments
+                            // parseResult.GetSuggestions does not work with textToMatch, only position
+                            // discuss options here
+                            //foreach (string? suggestion in command.GetSuggestions(templateParseResult, textToMatch))
+                            foreach (string? suggestion in templateParseResult.GetSuggestions())
```

Comment:
there is a dilemma here:
- `command.GetSuggestions(templateParseResult, textToMatch)` doesn't work for argument completion
- `templateParseResult.GetSuggestions()` only completes end of the line. 

We need mixture of those:
- the method should determine the symbol to complete (like `ParseResult.GetSuggestions`)
- the method should consider `textToMatch` like (`Command.GetSuggestions`)

Reply:
Correct way is to use `ParseResult.GetSuggestions`.  There should an API allowing to get position and raw input.

Reply:
https://github.com/dotnet/command-line-api/issues/1497